draft-ietf-tls-esni.md

```diff
@@ -347,14 +353,48 @@ For servers (in EncryptedExtensions), this extension contains the following
 structure:
 
 ~~~
+   enum {
+       esni_accept(0),
+       esni_retry_request(1),
+   } ServerESNIResponseType;
+
    struct {
-       uint8 nonce[16];
+       ServerESNIResponseType response_type;
+       select (response_type) {
+           case esni_accept:        uint8 nonce[16];
+           case esni_retry_request: ESNIKeys retry_keys<1..2^16-1>;
+       }
    } ServerEncryptedSNI;
 ~~~
 
+response_type
+: Indicates whether the server processed the client ESNI extension. (See
+{{client-behavior}} and {{server-behavior}}.}
+
 nonce
 : The contents of ClientESNIInner.nonce. (See {{client-behavior}}.)
 
+retry_keys
+: ESNIKeys structures containing the keys that the client should use on
+subsequent connections to encrypt the ClientESNIInner structure.
+
+This protocol also defines the "esni_required" alert, which is sent by the
+client when it offered an "encrypted_server_name" extension which was not
+accepted by the server.
+
+~~~~
+   enum {
+       esni_required(121),
+   } AlertDescription;
+~~~~
+
+Finally, implementations MUST track whether each PSK was established by a
```

Comment:
I'm confused.  This is the first mention of a PSK.  This needs more context.

The point here is that if you get a PSK from a previous connection (presumably an external PSK can have its own rules), then you need to ensure that you mark that PSK as being ESNI or not.  If it is ESNI, then you never resume from an ESNI PSK without ESNI and you never resume from a non-ESNI PSK if you are adding ESNI.

Without context, this is impossible to decipher.  I'm not even sure if I agree with the latter part of the requirement, so explaining why is probably necessary.

Also, this sentence is a run-on, so I can't really parse it.